### PR TITLE
Improve inference of anyFailAllFail

### PR DESCRIPTION
**Pull request by danking**
The core issue is that we want Scala to infer the T parameter
but not the C[_] parameter. Unfortunately, Scala does not provide
a facility to partially specify type parameters of a method
the changes herein achieve this with no run-time allocation
one cast, and one level of indirection.

### CODE REVIEW

File: src/main/scala/is/hail/utils/package.scala [hunk 1]
```diff
@@ -249,17 +249,23 @@ package object utils extends Logging
...
+      }
+      Some(b.result())
     }
-    Some(b.result())
   }
 
+  private final object anyFailAllFailInstance extends AnyFailAllFail[Nothing]
+
+  def anyFailAllFail[C[_]]: AnyFailAllFail[C] =
+    anyFailAllFailInstance.asInstanceOf[AnyFailAllFail[C]]
```
Comment (cseed) [inline src/main/scala/is/hail/utils/package.scala]: What a hack!  Please, let me get off this Scala train.  But, looks good to me.

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 2]
```diff
@@ -87,7 +87,7 @@ object FunctionRegistry {
             None
         }
 
-        anyFailAllFail[Array, Option[(Int, Transformation[Any, Any])]](conversions)
```
Comment (cseed) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: Are there other places it is used that  you can clean up (like the change I pushed yesterday)?

**Comment by danking** [comment]
@cseed I changed all of them I could find.

**Comment by cseed** [comment]
Needs rebase.

**Pull request merged**